### DISCUSSION TITLE
Dynamic Plugin StelGui problem

### DISCUSSION THREAD

**Discussion by emiemme** [Stellarium/stellarium#2382] [Q&A]
Hi everyone!

I'm trying to build a dynamic plugin.
I followed the stellarium-dynamic-plugin build instructions, everything compile and I can see the plugin in stellarium.
Now i want to add a button in the stellarium bottom toolbar.
I wrote a function in order to retrive the StelGui and add this button, just like the others plugins do. 

    void DynamicPluginTemplate::addButton()
    {
        StelGui* gui = dynamic_cast<StelGui*>(StelApp::getInstance().getGui());
        if (gui) {
            toolbarButton = new StelButton(Q_NULLPTR,
                               QPixmap("add_png"),
                               QPixmap("add_png"),
                               QPixmap("add_png"),
                               "fixme",
                               false,
                               "fixme");
            gui->getButtonBar()->addButton(toolbarButton, "fixme");
        } else {
            qDebug()<<"StelGui Dynamic cast failed...";
        }
    }
    /*************************************************************************
     Init our module
    *************************************************************************/
    void DynamicPluginTemplate::init()
    {
        qDebug() << "init called for DynamicPluginTemplate";
        addButton();
    }

I can still see the plugin in stellarium but the Dynamic cast always fails.
Is this the correct way to gain the access to StelGui?
What I'm missing?

**Comment by alex-w** [comment]
Wrong group of buttons and missing adding the icon into resources file. Check any static plugin with button on the bottom bar for example.

**Reply by emiemme** [reply]
Thank you for the reply.

Now I took the code from the calendars  plugin.
The addButton function is this:
```
void DynamicPluginTemplate::addButton()
{
    // Add a toolbar button
    StelApp& app=StelApp::getInstance();
    try
    {
        StelGui* gui = dynamic_cast<StelGui*>(app.getGui());
        if (gui!=Q_NULLPTR)
        {
            //qDebug() << "button...";
            toolbarButton = new StelButton(Q_NULLPTR,
                               QPixmap(":/DynamicPluginTemplate/bt_Calendars_On.png"),
                               QPixmap(":/DynamicPluginTemplate/bt_Calendars_Off.png"),
                               QPixmap(":/DynamicPluginTemplate/miscGlow32x32.png"),
                               "actionShow_DynamicPluginTemplate",
                               false,
                               "actionShow_DynamicPluginTemplate_dialog");
            //qDebug() << "add button...";
            gui->getButtonBar()->addButton(toolbarButton, "065-pluginsGroup");
        } else {
            qDebug()<<"StelGui Dynamic cast failed...";
        }
    }
    catch (std::runtime_error& e)
    {
        qWarning() << "WARNING: unable to create toolbar button for DynamicPluginTemplate plugin: " << e.what();
    }
}
```

The problem persist. The StelGui pointer is always NULL.

**Comment by emiemme** [comment]
UPDATE:

I tried to compile a static plugin with the same code of my dynamic plugin (I only changed the CMAKE configuration) and it seems to work.

Two questions: 
Is this code working for a dynamic plugin?
                                  `StelGui* gui = dynamic_cast<StelGui*>(StelApp::getInstance().getGui());`

Are there any examples of a dynamic plugin that use this code to add interface elements?

**Reply by alex-w** [reply]
Did you added all required includes?

**Reply by emiemme** [reply]
Yes, it compiles without errors,  I can activate the plugin inside stellarium but there is no pointer to StelGui.
I can share with you the code or at least the diff between my code and the dynamic plugin repo, but the diff are few lines of it.

**Reply by mel-co** [reply]
Are there any updates or workarounds? I wanted to hide the GUI and certain GUI elements.
Using ```StelMainScriptAPI::setGuiVisible(bool b)``` to hide the GUI works fine.
But when using ```StelApp::getInstance().getGui()->setVisible(b);``` directly in the dynamic plugin it does not work.
Also I have the same issue with ```StelGui* gui = dynamic_cast<StelGui*>(StelApp::getInstance().getGui());```, which results in no pointer to StelGui.
The plugin compiles without errors as well.

**Reply by 10110111** [reply]
Where do you call `DynamicPluginTemplate::addButton` from?

**Reply by mel-co** [reply]
I don't call addButton, I want to access StelGui to hide parts of the Gui.
But I have the same problem as mentioned here, where I am not able to get a pointer to StelGui from the dynamic Plugin.
But I called ```StelApp::getInstance().getGui()->setVisible(b);``` and
```StelGui* gui = dynamic_cast<StelGui*>(StelApp::getInstance().getGui());``` in the DynamicPluginTemplate.cpp, in a new public slot I added. (With the first one not working and the second resulting in an null_ptr)
I used the stellarium-dynamic-plugin project as a base and added the slot with the examples above, as well as a button in the dialog window, which is connected to the mentioned slot.

**Reply by alex-w** [reply]
@mel-co could you share your code?

**Reply by mel-co** [reply]
## MRE

Build and compiled using Qt6 and runs in stellarium.

### DynamicPluginTemplateWindow.cpp

Just changed the following function:

```C++
void DynamicPluginTemplateWindow::createDialogContent()
{
	ui->setupUi(dialog);

	dynamicPluginTemplate = GETSTELMODULE(DynamicPluginTemplate);

	connect(&StelApp::getInstance(), SIGNAL(languageChanged()), this, SLOT(retranslate()));
	connect(ui->closeStelWindow, SIGNAL(clicked()), this, SLOT(close()));
	connect(ui->startButton, SIGNAL(clicked()), dynamicPluginTemplate, SLOT(setup()));
}
```

### DynamicPluginTemplate.hpp

Only added the public slot

```C++
class DynamicPluginTemplate : public StelModule
{
	Q_OBJECT
public:
	...

public slots:
	virtual void setup();

private:
	...
};
```

### DynamicPluginTemplate.cpp

Added includes and the following definition

```C++
...

//added includes
#include "StelGuiBase.hpp"
#include "StelGui.hpp"

...

void DynamicPluginTemplate::setup(){

	StelApp::getInstance().getGui()->setVisible(false); //Not working

	StelGui* gui = dynamic_cast<StelGui*>(StelApp::getInstance().getGui()); //returns null_ptr

	if(gui == Q_NULLPTR){
		qInfo() << "gui is a nullptr";
	} else {
		qInfo() << "gui is not a nullptr";
	}
}
```

### dynamicPluginTemplateWindow.ui

Only added the button to ui

```xml
 ...
 <widget class="QWidget" name="dynamicPluginTemplateWindowForm">
  ...
  <widget class="QPushButton" name="startButton">
   <property name="geometry">
    <rect>
     <x>110</x>
     <y>60</y>
     <width>221</width>
     <height>51</height>
    </rect>
   </property>
   <property name="text">
    <string>START</string>
   </property>
  </widget>
 </widget>
 ...
```

Everything else is the same as in the [dynamic plugin example](https://github.com/Stellarium/stellarium-dynamic-plugin.git)

**Reply by 10110111** [reply]
Apparently, RTTI between shared libraries and the Stellarium executable is broken. My observation is that building Stellarium in Debug mode "fixes" this failure of `dynamic_cast`.
Your problem with hiding the GUI also only reproduces in Release mode, which is even more puzzling (and this may mean that RTTI is not the culprit).
